posts/2022-08-11-Rust-1.63.0.md

```diff
@@ -0,0 +1,233 @@
+currently must have ownership of any arguments passed into their closure; you
+can't pass borrowed data into a thread. In cases where the threads are expected
+to exit by the end of the function (by being `join()`'d), this isn't strictly
+necessary and can require workarounds like placing the data in an [`Arc`].
+
+Now, with 1.63.0, the standard library is adding *scoped* threads, which allow
+spawning a thread borrowing from the local stack frame. The
+[`std::thread::scope`] API provides the necessary guarantee that any spawned threads
+will have exited prior to itself returning, which allows for safely borrowing
+data. Here's an example:
+
+```rust
+let mut a = vec![1, 2, 3];
+let mut x = 0;
+
+std::thread::scope(|s| {
+    s.spawn(|| {
+        println!("hello from the first scoped thread");
+        // We can borrow `a` here.
+        dbg!(&a);
+    });
+    s.spawn(|| {
+        println!("hello from the second scoped thread");
+        // We can even mutably borrow `x` here,
+        // because no other threads are using it.
+        x += a[0] + a[2];
+    });
+    println!("hello from the main thread");
+});
+
+// After the scope, we can modify and access our variables again:
+a.push(4);
+assert_eq!(x, a.len());
+```
+
+[`std::thread::scope`]: https://doc.rust-lang.org/stable/std/thread/fn.scope.html
+[`std::thread::spawn`]: https://doc.rust-lang.org/stable/std/thread/fn.spawn.html
+[`Arc`]: https://doc.rust-lang.org/stable/std/sync/struct.Arc.html
+
+
+### Rust ownership for raw file descriptors/handles (I/O Safety)
+
+Previously, Rust code working with platform APIs taking raw file descriptors (on
+unix-style platforms) or handles (on Windows) would typically work directly with
+a platform-specific representation of the descriptor (for example, a `c_int`, or the alias `RawFd`).
+For Rust bindings to such native APIs, the type system then failed to encode
+whether the API would take ownership of the file descriptor (e.g., `close`) or
+merely borrow it (e.g., `dup`).
+
+Now, Rust provides wrapper types such as [`BorrowedFd`] and [`OwnedFd`], which are marked as
+`#[repr(transparent)]`, meaning that `extern "C"` bindings can directly take
+these types to encode the ownership semantics. See the stabilized APIs section
+for the full list of wrapper types stabilized in 1.63, currently, they are
+available on cfg(unix) platforms, Windows, and WASI.
+
+We recommend that new APIs use these types instead of the previous type aliases
+(like [`RawFd`]).
+
+[`RawFd`]: https://doc.rust-lang.org/stable/std/os/unix/io/type.RawFd.html
+[`BorrowedFd`]: https://doc.rust-lang.org/stable/std/os/unix/io/struct.BorrowedFd.html
+[`OwnedFd`]: https://doc.rust-lang.org/stable/std/os/unix/io/struct.OwnedFd.html
+
+### Turbofish for generics in functions with `impl Trait`
+
+The turbofish notation (`foo::<u32>()`) can be used to specify the types for explicit type
+parameters (`fn foo<T: Copy>`) even when `impl Trait` is used in the argument types.
```

Comment:
FWIW: I've tried to parse this many times and I still don't understand it.... Could this be worded differently to make it clear to folks like me?